modules/fedimint-mint/src/lib.rs

```diff
@@ -103,6 +117,154 @@ pub struct VerificationCache {
+        &self,
+        peers: &[PeerId],
+        params: &ModuleConfigGenParams,
+    ) -> (BTreeMap<PeerId, ServerModuleConfig>, ClientModuleConfig) {
+        let tbs_keys = params
+            .mint_amounts
+            .iter()
+            .map(|&amount| {
+                let (tbs_pk, tbs_pks, tbs_sks) = dealer_keygen(peers.threshold(), peers.len());
+                (amount, (tbs_pk, tbs_pks, tbs_sks))
+            })
+            .collect::<HashMap<_, _>>();
+
+        let mint_cfg: BTreeMap<_, MintConfig> = peers
+            .iter()
+            .map(|&peer| {
+                let config = MintConfig {
+                    threshold: peers.threshold(),
+                    tbs_sks: params
+                        .mint_amounts
+                        .iter()
+                        .map(|amount| (*amount, tbs_keys[amount].2[peer.to_usize()]))
+                        .collect(),
+                    peer_tbs_pks: peers
+                        .iter()
+                        .map(|&key_peer| {
+                            let keys = params
+                                .mint_amounts
+                                .iter()
+                                .map(|amount| (*amount, tbs_keys[amount].1[key_peer.to_usize()]))
+                                .collect();
+                            (key_peer, keys)
+                        })
+                        .collect(),
+                    fee_consensus: FeeConsensus::default(),
+                };
+                (peer, config)
+            })
+            .collect();
+
+        let client_cfg = MintClientConfig {
+            tbs_pks: tbs_keys
+                .into_iter()
+                .map(|(amount, (pk, _, _))| (amount, pk))
+                .collect(),
+            fee_consensus: FeeConsensus::default(),
+        };
+
+        (
+            mint_cfg
+                .into_iter()
+                .map(|(k, v)| {
+                    (
+                        k,
+                        serde_json::to_value(v)
+                            .expect("serialization can't fail")
+                            .into(),
+                    )
+                })
+                .collect(),
+            serde_json::to_value(client_cfg)
+                .expect("serialization can't fail")
+                .into(),
+        )
+    }
+
+    async fn distributed_gen(
+        &self,
+        connections: &mut AnyPeerConnections<ConfigGenPeerMsg>,
+        our_id: &PeerId,
+        peers: &[PeerId],
+        params: &ModuleConfigGenParams,
+        _task_group: &mut TaskGroup,
+    ) -> anyhow::Result<Cancellable<(ServerModuleConfig, ClientModuleConfig)>> {
+        let mut dkg = DkgRunner::multi(
+            params.mint_amounts.to_vec(),
+            peers.threshold(),
+            our_id,
+            peers,
+        );
+        let g2 = if let Ok(g2) = dkg.run_g2(connections, &mut OsRng).await {
+            g2
+        } else {
+            return Ok(Err(Cancelled));
```

Comment:
In what circumstance does this happen?

Reply:
E.g. user pressed ctrl+c or another thread failed so the whole program should gracefully shutdown as well.